torch/testing/_comparison.py

```diff
@@ -586,10 +587,29 @@ def _check_supported(self, tensor: torch.Tensor, *, id: Tuple[Any, ...]) -> None
     def compare(self) -> None:
         actual, expected = self.actual, self.expected
 
-        self._compare_attributes(actual, expected)
-        actual, expected = self._equalize_attributes(actual, expected)
+        with self._handle_meta_tensor_data_access():
+            self._compare_attributes(actual, expected)
+            actual, expected = self._equalize_attributes(actual, expected)
 
-        self._compare_values(actual, expected)
+            self._compare_values(actual, expected)
+
+    @contextlib.contextmanager
+    def _handle_meta_tensor_data_access(self):
+        """Turns a vanilla :class:`NotImplementedError` stemming from data access on a meta tensor into an expressive
+        :class:`ErrorMeta`.
+
+        Although it looks like meta tensors could be handled upfront, we need to do it lazily: there are use cases
+        where a meta tensor wraps a data tensors and dispatches all operator calls to it. Thus, although the tensor is
+        a meta tensor, it behaves like a regular one.
+        """
+        try:
+            yield
+        except NotImplementedError as error:
+            if "meta" not in str(error).lower():
```

Comment:
I'm not in love with this mechanism for detecting if the not impl error is based on the tensor being a meta tensor or not, but it seems OK for now

Reply:
Me neither. But I don't think there is a better option right now. We should remove this while resolving #68592.